### PR TITLE
App server logging

### PR DESCRIPTION
**Pull request by illume**
# App server logging

Log errors when something goes wrong executing or running the server process in the app.
With this PR the app logs stderr, and stdout from the `server` process.
Additionally the signal, and return code of the `server` process are logged.

related: https://github.com/kinvolk/headlamp/issues/112

What's not in this PR:
- updating the UI when there's a `server` related error.
- restarting or retrying server process start

Should we be logging the server.stderr URLs it prints out?

# How to use

1. log when server process crashes or is killed.

```bash
make app-mac
```

run with logging enabled.
```bash
app/dist/mac/Headlamp.app/Contents/MacOS/Headlamp --enable-logging
```

In a separate terminal you can kill the `server` process.
```bash
while true; do killall server ; done
```

Now you should see an error log that the process terminated.

Screenshot 2021-01-11 at 09 55 56

2. When running, and quitting without the server process it should not log an error.

Screenshot 2021-01-11 at 09 56 49

3. remove the server process.

To replicate an issue executing the server process, remove the file.

```bash
rm app/dist/mac/Headlamp.app/Contents//Resources/electron/server
```

Then an error should be logged.

Screenshot 2021-01-11 at 09 45 52

# Testing done

See above on the testing done on Mac.

### CODE REVIEW

File: app/electron/main.js [hunk 1]
```diff
@@ -123,11 +123,37 @@ const isDev = process.mainModule.filename.indexOf('app.asar') === -1 ||
...
+    serverProcess.on('close', (code, signal) => {
+      const closeMessage = `server process process exited with code:${code} signal:${signal}`;
+      if (!intentionalQuit) {
+        //TODO: message mainWindow, or loadURL to an error url?
+        log.error(closeMessage);
+      } else {
+        log.info(closeMessage);
+      }
+      serverProcessQuit = true;
+    });
```
Comment (ashu8912) [inline app/electron/main.js:159]: @illume can we move these event handlers on serverProcess into a seperate function maybe something like ```attachServerEventHandlers```(which resides in the global scope of main.js) because otherwise than in the headless mode feature  https://github.com/kinvolk/headlamp/pull/154 I will have to copy the entire server handlers again

Comment (illume) [reply app/electron/main.js]: Good idea. Thanks @ashu8912 

I pushed a commit for this.

File: app/electron/main.js [hunk 2]
```diff
@@ -170,8 +196,16 @@ app.on('ready', createWindow);
...
   if (serverProcess) {
-    serverProcess.stdin.pause();
-    serverProcess.kill();
+    if (serverProcessQuit) {
+      log.error('server process already not running');
+    } else {
+      intentionalQuit = true;
+      log.info('stopping server process...');
+      serverProcess.stdin.pause();
+      // TODO: should we try and end the process a bit more gracefully?
```
Comment (joaquimrocha) [inline app/electron/main.js:215]: We've been using `// @todo: `. Not too important but maybe we keep it consistent so it's easy to find in the future.
Maybe the phrasing could also be styled as a todo instead of a question.

File: app/electron/main.js [hunk 3]
```diff
@@ -126,37 +126,44 @@ let serverProcess = null;
 let intentionalQuit = false;
 let serverProcessQuit = false;
 
+/**
+ * add some error handlers to the serverProcess.
+ * @param  {ChildProcess} serverProcess to attach the error handlers to.
+ */
+const attachServerEventHandlers = (serverProcess) => {
```
Comment (joaquimrocha) [inline app/electron/main.js:133]: We've been using `function ...` instead of the arrow functions.
Maybe we should move to arrows functions but for consistency for now, let's keep the functions.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]

File: app/electron/main.js (hunk 2 above)
Comment (illume) [reply app/electron/main.js]: Ok, thanks. It's done.

File: app/electron/main.js (hunk 3 above)
Comment (illume) [reply app/electron/main.js]: Ok, thanks. It's done.

File: app/electron/main.js [hunk 4]
```diff
@@ -205,7 +212,7 @@ app.on('quit', function () {
       intentionalQuit = true;
       log.info('stopping server process...');
       serverProcess.stdin.pause();
-      // TODO: should we try and end the process a bit more gracefully?
+      // @todo: should we try and end the process a bit more gracefully?
```
Comment (joaquimrocha) [inline app/electron/main.js:215]: This doesn't belong in this commit. Please add it to the commit that introduced the todo.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Left just a comment in the code. Also please remove the periods on commit titles.

File: app/electron/main.js (hunk 4 above)
Comment (illume) [reply app/electron/main.js:215]: I squished the commits into one now. Can you please check it again?

**Comment by illume** [comment]
I have one open question left...
> Should we be logging the server.stderr URLs it prints out?

Here's an example:

Screenshot 2021-01-11 at 13 57 34

**Comment by joaquimrocha** [comment]
> I have one open question left...
> 
> > Should we be logging the server.stderr URLs it prints out?
> 
> Here's an example:
> 
> Screenshot 2021-01-11 at 13 57 34

I think so. If it gets to be too much, we can silence it through levels in the future.

**Review by joaquimrocha — APPROVED** [review]
